crates/sui-core/src/consensus_throughput_calculator.rs

```diff
@@ -0,0 +1,620 @@
+/// throughput as we need to make sure that validators will see an as possible consistent view to assign
+/// the right profile.
+pub struct ConsensusThroughputProfiler {
+    /// The throughput profile will be eligible for update every `throughput_profile_update_interval` seconds.
+    /// A bucketing approach is followed where the throughput timestamp is used in order to calculate on which
+    /// seconds bucket is assigned to. When we detect a change on that bucket then an update is triggered (if a different
+    /// profile is calculated). That allows validators to align on the update timing and ensure they will eventually
+    /// converge as the consensus timestamps are used.
+    throughput_profile_update_interval: TimestampSecs,
+    /// When current calculated throughput (A) is lower than previous, and the assessed profile is now a lower than previous,
+    /// we'll change to the lower profile only when (A) <= (previous_profile.throughput) * (100 - throughput_profile_cool_down_threshold) / 100.
+    /// Otherwise we'll stick to the previous profile. We want to do that to avoid any jittery behaviour that alternates between two profiles.
+    throughput_profile_cool_down_threshold: u64,
+    /// The profile ranges to use to profile the throughput
+    profile_ranges: ThroughputProfileRanges,
+    /// The most recently calculated throughput profile
+    last_throughput_profile: ArcSwap<ThroughputProfileEntry>,
+    metrics: Arc<AuthorityMetrics>,
+    /// The throughput calculator to use to derive the current throughput.
+    calculator: Arc<ConsensusThroughputCalculator>,
+}
+
+impl ConsensusThroughputProfiler {
+    pub fn new(
+        calculator: Arc<ConsensusThroughputCalculator>,
+        throughput_profile_update_interval: Option<TimestampSecs>,
+        throughput_profile_cool_down_threshold: Option<u64>,
+        metrics: Arc<AuthorityMetrics>,
+        profile_ranges: ThroughputProfileRanges,
+    ) -> Self {
+        let throughput_profile_update_interval = throughput_profile_update_interval
+            .unwrap_or(DEFAULT_THROUGHPUT_PROFILE_UPDATE_INTERVAL_SECS);
+        let throughput_profile_cool_down_threshold = throughput_profile_cool_down_threshold
+            .unwrap_or(DEFAULT_THROUGHPUT_PROFILE_COOL_DOWN_THRESHOLD);
+
+        assert!(
+            throughput_profile_update_interval > 0,
+            "throughput_profile_update_interval should be >= 0"
+        );
+
+        assert!(
+            (0..=30).contains(&throughput_profile_cool_down_threshold),
+            "Out of bounds provided cool down threshold offset"
+        );
+
+        Self {
+            throughput_profile_update_interval,
+            throughput_profile_cool_down_threshold,
+            profile_ranges: Default::default(),
+            last_throughput_profile: ArcSwap::from_pointee(ThroughputProfileEntry {
+                profile: profile_ranges.lowest_profile(),
```

Comment:
Btw, we need to choose the default behavior carefully, or maybe persist the profile in store. The default profile should not cause amplified submissions. Otherwise when the network has higher load and a fraction of validators restart, the load will be even more amplified.

Reply:
Agreed - initially had this to higher profile to be more proactive and I'll probably revert it.